### PR TITLE
[Finetune] Fix evaluation

### PR DESCRIPTION
**Pull request by minmingzhu**
After trying several solutions, I finally used the tokenize_func from neural_chat:
1. Using the transformer tokenizer with padding and attention_mask, and copying input_ids to labels.
2. Using the transformer tokenizer with padding and attention_mask, and masking the input in the prompt.
3. Using the transformer tokenizer with padding and attention_mask, and masking the response in the prompt.
4. Using neural_chat's tokenize_func for padding, filling attention_mask and labels which masking the input in the prompt.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/minmingz/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/minmingz/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Case | Average | ARC | HellaSwag | MMLU | TruthfulQA | Winogrande
-- | -- | -- | -- | -- | -- | --
Case 1 | 62.116 | 55.97 | 78.39 | 60.75 | 45.86 | 69.61
Case 2 | 58.99 | 54.18 | 72.85 | 60.23 | 42.18 | 65.51
Case 3 | 59.064 | 53.24 | 73.21 | 60.85 | 41.56 | 66.46
Case 4 | 63.03 | 53.67 | 81.47 | 60.27 | 43.02 | 76.72

</body>

</html>

### CODE REVIEW

File: llm_on_ray/finetune/DataPreprocess.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
+#
+# Copyright 2023 The LLM-on-Ray Authors.
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:2]: rename this file: data_preprocess.py

File: llm_on_ray/finetune/finetune.py [hunk 2]
```diff
@@ -37,10 +40,12 @@
 from pydantic_yaml import parse_yaml_raw_as
 
 from llm_on_ray import common
-from llm_on_ray.finetune import template
+from llm_on_ray.finetune.DataPreprocess import AlpacaDataPreprocess
 from llm_on_ray.finetune.finetune_config import FinetuneConfig
 from importlib import util
 
+IGNORE_INDEX = -100
```
Comment (harborn) [inline llm_on_ray/finetune/finetune.py:47]: seems this variable not used

File: llm_on_ray/finetune/DataPreprocess.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+import copy
+import re
+from itertools import chain
+
+import torch
+
+IGNORE_INDEX = -100
+
+
+class AlpacaDataPreprocess:
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:26]: does this class only used for preprocessing alpaca dataset?

File: llm_on_ray/finetune/DataPreprocess.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                    + self.end
+                    + "\n"
+                    + self.response
+                )
+                prompts["prompt_sources"].append(prompt)
+            prompt_response = response + self.end
+            prompts["prompt_targets"].append(prompt_response)
+        return prompts
+
+    def tokenize_func(self, tokenizer, config):
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:76]: rename this function:
`def tokenize`

File: llm_on_ray/finetune/DataPreprocess.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import re
+from itertools import chain
+
+import torch
+
+IGNORE_INDEX = -100
+
+
+class AlpacaDataPreprocess:
+    def __init__(self, eos_token):
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:27]: pass config here:
`def __init__(self, config):`

File: llm_on_ray/finetune/DataPreprocess.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                    + "\n"
+                    + self.response
+                )
+                prompts["prompt_sources"].append(prompt)
+            prompt_response = response + self.end
+            prompts["prompt_targets"].append(prompt_response)
+        return prompts
+
+    def tokenize_func(self, tokenizer, config):
+        padding_side = config["Dataset"].get("padding_side", "right")
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:77]: not pass config in this function, pass config in class construction

File: llm_on_ray/finetune/DataPreprocess.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                assert len(prompt_ids) <= max_source_length
+                assert len(labels) == len(input_ids) == len(attention_mask)
+
+                examples["input_ids"].append(torch.tensor(input_ids))
+                examples["labels"].append(labels)
+                examples["attention_mask"].append(attention_mask)
+
+            return examples
+
+        def preprocess_function_with_tokenizer(examples):
```
Comment (harborn) [inline llm_on_ray/finetune/DataPreprocess.py:183]: this function is not clear too much compared with `preprocess_function_with_tokenize`
will get confused from function name

File: llm_on_ray/finetune/finetune.py (hunk 2 above)
Comment (minmingzhu) [reply llm_on_ray/finetune/finetune.py:47]: remove

File: llm_on_ray/finetune/DataPreprocess.py (hunk 3 above)
Comment (minmingzhu) [reply llm_on_ray/finetune/DataPreprocess.py:26]: Yes, In this PR, it just preprocess alpaca dataset. But in DPO PR, it will be add DPOIntelOrcaPreprocesser class to preprocess intel orca dataset.

File: llm_on_ray/finetune/data_preprocess.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+import copy
+import re
+from itertools import chain
+
+import torch
+
+IGNORE_INDEX = -100
+
+
+class DataPreprocess:
```
Comment (harborn) [inline llm_on_ray/finetune/data_preprocess.py:26]: suggest to split this class two parts:
1. make prompt -> data_temmplate.py
2. tokenize -> data_process.py

File: llm_on_ray/finetune/data_preprocess.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    # We used the following prompts for fine-tuning the Alpaca model. You can find reference doc form this URL(https://github.com/tatsu-lab/stanford_alpaca/blob/main/README.md#data-release)
+    def __init__(self, config, eos_token):
+        self.config = config
+        self.end = eos_token
+        self.intro = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
+        self.instruction = "### Instruction:\n"
+        self.input = "### Input:\n"
+        self.response = "### Response:\n"
+
+    def prompt(self, examples):
```
[the remark this answers is not in the document]
Comment (minmingzhu) [reply llm_on_ray/finetune/data_preprocess.py]: I think that if we add new dataset, we just override make_prompt function to process data. If set this as a global function, Once there is a new data set that needs to be processed, it cannot be reused.

File: llm_on_ray/finetune/data_preprocess.py (hunk 8 above)
Comment (minmingzhu) [reply llm_on_ray/finetune/data_preprocess.py]: I have rename python file name and function name. But I still put the make_prompt and tokenzie function in a class. Beacuse both need to use common variables.

image

File: llm_on_ray/finetune/data_process.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+                    + self.end
+                    + "\n"
+                    + self.response
+                )
+                prompts["prompt_sources"].append(prompt)
+            prompt_response = response + self.end
+            prompts["prompt_targets"].append(prompt_response)
+        return prompts
+
+    def tokenize(self, tokenizer):
```
Comment (harborn) [inline llm_on_ray/finetune/data_process.py:78]: this function do nothing, just provide a sub function.
so I think you should put following variable to class's `__init__` function, and provide following two functions:
1. `def tokenize_by_nerual_chat(examples):`
2. `def tokenize(examples):`

**Review by harborn — APPROVED** [review]
